### PR TITLE
Support for multiple Source languages

### PR DESCRIPTION
**Pull request by cjw296**
Initially widening to ReST, MyST and Python source files.

### CODE REVIEW

File: tests/functional/myst/doctest.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# DocTest examples
+
+A Python REPL / DocTest example using normal way of specifying Python
+language:
+
+```python
+>>> 1+1
+2
+```
```
Comment (cjw296) [inline tests/functional/myst/doctest.md:9]: @hynek - is this how you do doctest-style Python REPL stuff in MyST or is there a better way?

Comment (hynek) [reply tests/functional/myst/doctest.md:9]: sorry for the delay…the current way to do doctest in Sphinx + MyST looks like this:

    ```{eval-rst}
    .. doctest::

       >>> from structlog import get_logger
       >>> from structlog.testing import capture_logs
       >>> with capture_logs() as cap_logs:
       ...    get_logger().bind(x="y").info("hello")
       >>> cap_logs
       [{'x': 'y', 'event': 'hello', 'log_level': 'info'}]
    ```

So if you implement that, you'd be drop-in.

---

I hope you might be able to work with the MyST devs to give us

    ```{doctest}

or

    ```{sybil}

for that matter?

Comment (cjw296) [reply tests/functional/myst/doctest.md:9]: If you try:

    ```python
       >>> from structlog import get_logger
       >>> from structlog.testing import capture_logs
       >>> with capture_logs() as cap_logs:
       ...    get_logger().bind(x="y").info("hello")
       >>> cap_logs
       [{'x': 'y', 'event': 'hello', 'log_level': 'info'}]
    ```

I think you'll find the pygments rendering does the right thing, so I'll have sybil support that.

If you use:

    ```{eval-rst}
    .. doctest::

       >>> from structlog import get_logger
       >>> from structlog.testing import capture_logs
       >>> with capture_logs() as cap_logs:
       ...    get_logger().bind(x="y").info("hello")
       >>> cap_logs
       [{'x': 'y', 'event': 'hello', 'log_level': 'info'}]
    ```

...then you won't have to change anything, but just use Sybil's resting `sybil.parsers.rest.DocTestParser` to pick them out.

I'm afraid I don't have bandwidth to contribute to MyST, but if they add a `{doctest}` role, then adding support to Sybil will be trivial.

Comment (hynek) [reply tests/functional/myst/doctest.md:9]: By "work" I meant "ask them for it". :) I think they'd be happy to get rid of every instance of `{eval-rst}`

Comment (cjw296) [reply tests/functional/myst/doctest.md:9]: Why wouldn't you use the python-code-block-with-repl-in-it which works already?

        A Python REPL / DocTest example using normal way of specifying Python
        language:
        
        ```python
        >>> 1+1
        2
        ```

...renders as:
Screenshot 2022-11-01 at 13 25 42

The "doctest" role in ReST looks like a thing invented by `sphinx.ext.doctest`, which isn't needed for Sybil (although stuff written in that style will "just work"), so I'd expect the MyST folks to say it isn't needed for them either.

Comment (hynek) [reply tests/functional/myst/doctest.md:9]: That's true but I have tons of code blocks like that that aren't valid/runnable doctests. Assuming every Python code block that looks like a session will break A LOT of builds.

Comment (cjw296) [reply tests/functional/myst/doctest.md:9]: Can you point me at some examples of those? It'd be interesting to think about the best way to handle them...

Comment (hynek) [reply tests/functional/myst/doctest.md:9]: On a phone rn but imagine someone wants to demonstrate how easy httpx is:

    ```Python
    >>> httpx.get("whatever")
    ...
    ```

The first problem is that you're making network calls. :)

Comment (cjw296) [reply tests/functional/myst/doctest.md:9]: Sure, but that's the same as anything like that which you want to test: you need to mock... here's how I did it for [configurator](https://configurator.readthedocs.io/en/latest/):

https://github.com/simplistix/configurator/blob/master/conftest.py
https://github.com/simplistix/configurator/blame/master/README.rst#L28-L36

For existing stuff, you can add [skips](https://sybil.readthedocs.io/en/latest/parsers.html#skip) (which I'll definitely be adding in the MyST support!).

That said, if you have some real-world existing examples, it'd be interesting to see if a custom parser or some parser option could be added to ease migration.

Comment (hynek) [reply tests/functional/myst/doctest.md:9]: I'm bearing great news! According to @Tinche

    ```{doctest}

already works!

c.f. https://github.com/python-attrs/cattrs/blob/tin/tagged-unions/docs/strategies.md

So how about being conservative and support only those two use-cases for now? 😇

**Comment by cjw296** [comment]
@hynek / @Tinche - how do you folks do multi-line comments in MyST? I've seen this style, but is there anything else?

```
% invisible-code-block: python
%
% b = 4
%
% ...etc...
```

**Comment by hynek** [comment]
> @hynek / @Tinche - how do you folks do multi-line comments in MyST? I've seen this style, but is there anything else?
> 
> ```
> % invisible-code-block: python
> %
> % b = 4
> %
> % ...etc...
> ```

No clue!

File: tests/functional/myst/python.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+Here's another way we might be able to do them:
+
+<!---  invisible-code-block: python
+def foo():
+   return 42
+
+meaning_of_life = 42
+
+assert foo() == meaning_of_life()
+--->
```
Comment (cjw296) [inline tests/functional/myst/python.md:33]: @hynek: looks like both of the above work 🎉

**Pull request merged**